aws/data_source_aws_instances.go

```diff
@@ -50,8 +50,14 @@ func dataSourceAwsInstancesRead(d *schema.ResourceData, meta interface{}) error
 	params := &ec2.DescribeInstancesInput{
 		Filters: []*ec2.Filter{
 			&ec2.Filter{
-				Name:   aws.String("instance-state-name"),
-				Values: []*string{aws.String("running")},
```

Comment:
The enhancement here needs to allow the `instance-state` to either be configurable, while leaving the existing default behavior to not break backwards compatibility. This can be accomplished two ways:

Less preferably as a string attribute, but provided as an example, e.g.

```go
"instance_state_name": {
  Type: schema.TypeString,
  Optional: true,
  Default: ec2.InstanceStateNameRunning,
  ValidateFunc: validation.StringInSlice([]string{
    ec2.InstanceStateNamePending,
    ec2.InstanceStateNameRunning,
    ec2.InstanceStateNameShuttingDown,
    ec2.InstanceStateNameStopped,
    ec2.InstanceStateNameStopping,
    ec2.InstanceStateNameTerminated,
  }, false),
},
```

Where it is referenced in the code with:

```go
	params := &ec2.DescribeInstancesInput{
		Filters: []*ec2.Filter{
			&ec2.Filter{
				Name: aws.String("instance-state-name"),
				Values: []*string{aws.String(d.Get("instance_state_name").(string))},
			},
		},
	}
```

Or, more preferably, as a set so you can configure one or more instance states:

```go
"instance_state_names": {
  Type: schema.TypeSet,
  Optional: true,
  Elem: schema.Schema{
    Type: schema.TypeString,
    ValidateFunc: validation.StringInSlice([]string{
      ec2.InstanceStateNamePending,
      ec2.InstanceStateNameRunning,
      ec2.InstanceStateNameShuttingDown,
      ec2.InstanceStateNameStopped,
      ec2.InstanceStateNameStopping,
      ec2.InstanceStateNameTerminated,
    }, false),
  },
},
```

Where it is referenced in the code with:

```go
	instanceStateNames := []*string{aws.String(ec2.InstanceStateNameRunning)}
	if v, ok := d.GetOk("instance_state_names"); ok && len(v.(*schema.Set).List()) > 0 {
		instanceStateNames = expandStringSet(v.(*schema.Set))
	}
	params := &ec2.DescribeInstancesInput{
		Filters: []*ec2.Filter{
			&ec2.Filter{
				Name: aws.String("instance-state-name"),
				Values: instanceStateNames,
			},
		},
	}
```

Either case should implement a new acceptance test that creates two instances and immediately calls the data source that accepts the `pending` (and probably `running` as well) instance states.